File: python/src/citenexus/extract/md.py
"""MdExtractor — Markdown headings (level + structure_path) and paragraphs (§8)."""

from __future__ import annotations

from typing import Any

from markdown_it import MarkdownIt

from citenexus.extract.plain import load_text
from citenexus.extract.types import (
    BlockKind,
    ExtractedBlock,
    ExtractedDoc,
    SourceType,
    StructureType,
)
from citenexus.plugins.base import ExtractorPlugin


class MdExtractor(ExtractorPlugin):
    """Headings become heading blocks (with a level + ancestor structure_path);
    paragraphs become paragraph blocks tagged with the enclosing heading path."""

    plugin_version = "md/1"

    def __init__(self, document_id: str | None = None) -> None:
        self.document_id = document_id

    def extract(self, source: Any) -> ExtractedDoc:
        text, doc_id, source_uri = load_text(source, self.document_id)
        tokens = MarkdownIt().parse(text)

        blocks: list[ExtractedBlock] = []
        stack: list[tuple[int, str]] = []  # (level, heading text) ancestors
        order = 0
        has_heading = False

        i = 0
        while i < len(tokens):
            tok = tokens[i]
            if tok.type == "heading_open":
                level = int(tok.tag[1:])
                content = tokens[i + 1].content.strip()
                while stack and stack[-1][0] >= level:
                    stack.pop()
                ancestors = tuple(t for _, t in stack)
                blocks.append(
                    ExtractedBlock(
                        order=order,
                        kind=BlockKind.heading,
                        text=content,
                        level=level,
                        structure_path=ancestors,
                    )
                )
                stack.append((level, content))
                has_heading = True
                order += 1
                i += 3
            elif tok.type == "paragraph_open":
                content = tokens[i + 1].content.strip()
                if content:
                    blocks.append(
                        ExtractedBlock(
                            order=order,
                            kind=BlockKind.paragraph,
                            text=content,
                            structure_path=tuple(t for _, t in stack),
                        )
                    )
                    order += 1
                i += 3
            elif tok.type in ("fence", "code_block"):
                # BlockKind.code / EUType.code_block are defined and mapped
                # (evidence/builder.py:_KIND_TO_TYPE) but no extractor ever
                # constructed one — a fenced (```lang) or 4-space-indented
                # code block is a single token (not an open/close pair), with
                # its literal source in .content (never HTML-escaped/inline
                # tokens like paragraph/heading text is).
                content = tok.content.rstrip("\n")
                if content:
                    blocks.append(
                        ExtractedBlock(
                            order=order,
                            kind=BlockKind.code,
                            text=content,
                            structure_path=tuple(t for _, t in stack),
                        )
                    )
                    order += 1
                i += 1
            else:
                i += 1

        structure = StructureType.heading_tree if has_heading else StructureType.none
        return ExtractedDoc(
            document_id=doc_id,
            source_type=SourceType.md,
            structure_type=structure,
            source_uri=source_uri,
            blocks=tuple(blocks),
        )

File: rust/src/extract/md.rs
//! MdExtractor — headings become heading blocks (level + ancestor
//! `structure_path`); paragraphs carry the enclosing heading path
//! (mirrors `extract/md.py`).

use pulldown_cmark::{Event, HeadingLevel, Parser, Tag, TagEnd};

use crate::types::*;

fn level_of(level: HeadingLevel) -> u32 {
    match level {
        HeadingLevel::H1 => 1,
        HeadingLevel::H2 => 2,
        HeadingLevel::H3 => 3,
        HeadingLevel::H4 => 4,
        HeadingLevel::H5 => 5,
        HeadingLevel::H6 => 6,
    }
}

pub fn extract(text: &str, document_id: &str, source_uri: Option<String>) -> ExtractedDoc {
    let mut blocks: Vec<ExtractedBlock> = Vec::new();
    let mut stack: Vec<(u32, String)> = Vec::new(); // (level, heading text)
    let mut order = 0usize;
    let mut has_heading = false;

    // Collector state: Some(kind) while inside a heading/paragraph.
    enum In {
        Heading(u32),
        Paragraph,
    }
    let mut current: Option<(In, String)> = None;

    for event in Parser::new(text) {
        match event {
            Event::Start(Tag::Heading { level, .. }) => {
                current = Some((In::Heading(level_of(level)), String::new()));
            }
            Event::Start(Tag::Paragraph) => {
                current = Some((In::Paragraph, String::new()));
            }
            Event::Text(t) | Event::Code(t) => {
                if let Some((_, buf)) = current.as_mut() {
                    buf.push_str(&t);
                }
            }
            Event::SoftBreak | Event::HardBreak => {
                if let Some((_, buf)) = current.as_mut() {
                    buf.push('\n');
                }
            }
            Event::End(TagEnd::Heading(_)) => {
                if let Some((In::Heading(level), buf)) = current.take() {
                    let content = buf.trim().to_string();
                    while stack.last().is_some_and(|(l, _)| *l >= level) {
                        stack.pop();
                    }
                    let ancestors: Vec<String> = stack.iter().map(|(_, t)| t.clone()).collect();
                    blocks.push(ExtractedBlock {
                        order,
                        kind: BlockKind::Heading,
                        text: content.clone(),
                        page: None,
                        bbox: None,
                        level: Some(level),
                        start_line: None,
                        end_line: None,
                        structure_path: ancestors,
                        cells: vec![],
                    });
                    stack.push((level, content));
                    has_heading = true;
                    order += 1;
                }
            }
            Event::End(TagEnd::Paragraph) => {
                if let Some((In::Paragraph, buf)) = current.take() {
                    let content = buf.trim().to_string();
                    if !content.is_empty() {
                        blocks.push(ExtractedBlock {
                            order,
                            kind: BlockKind::Paragraph,
                            text: content,
                            page: None,
                            bbox: None,
                            level: None,
                            start_line: None,
                            end_line: None,
                            structure_path: stack.iter().map(|(_, t)| t.clone()).collect(),
                            cells: vec![],
                        });
                        order += 1;
                    }
                }
            }
            _ => {}
        }
    }

    let structure = if has_heading {
        StructureType::HeadingTree
    } else {
        StructureType::None
    };
    ExtractedDoc {
        document_id: document_id.to_string(),
        source_type: SourceType::Md,
        structure_type: structure,
        source_uri,
        blocks,
        images: vec![],
    }
}
